### PR TITLE
New module: nanoq

### PR DESCRIPTION
**Pull request by LilyAnderssonLee**
## PR checklist
Related to https://github.com/nf-core/taxprofiler/issues/415

- [ ] This comment contains a description of changes (with reason).
- [ ] If you've fixed a bug or added code that should be tested, add tests!
- [ ] If you've added a new tool - have you followed the module conventions in the [contribution docs](https://github.com/nf-core/modules/tree/master/.github/CONTRIBUTING.md)
- [ ] If necessary, include test data in your PR.
- [ ] Remove all TODO statements.
- [ ] Emit the `versions.yml` file.
- [ ] Follow the naming conventions.
- [ ] Follow the parameters requirements.
- [ ] Follow the input/output options guidelines.
- [ ] Add a resource `label`
- [ ] Use BioConda and BioContainers if possible to fulfil software requirements.
- Ensure that the test works with either Docker / Singularity. Conda CI tests can be quite flaky:
  - For modules:
    - [ ] `nf-core modules test <MODULE> --profile docker`
    - [ ] `nf-core modules test <MODULE> --profile singularity`
    - [ ] `nf-core modules test <MODULE> --profile conda`
  - For subworkflows:
    - [ ] `nf-core subworkflows test <SUBWORKFLOW> --profile docker`
    - [ ] `nf-core subworkflows test <SUBWORKFLOW> --profile singularity`
    - [ ] `nf-core subworkflows test <SUBWORKFLOW> --profile conda`

### CODE REVIEW

File: modules/nf-core/nanoq/tests/main.nf.test.snap [hunk 1]
```diff
@@ -0,0 +1,307 @@
...
+                ],
+                "1": [
+                    [
+                        {
+                            "id": "test",
+                            "single_end": true
+                        },
+                        [
+                            "test.stats:md5,d41d8cd98f00b204e9800998ecf8427e",
+                            "/Users/lili.li.2/modules/.nf-test/tests/9d4f60427672ca3da40466350f6c67ae/work/08/8e3658c9a4909104a7edb60966757c/test_nanoq.fastq.gz",
```
Comment (SPPearce) [inline modules/nf-core/nanoq/tests/main.nf.test.snap:144]: There is an issue with nf-test and gzipped files. If you have a file ending .gz that isn't actually generated by gzip it gives this weirdness, where the filepath gets captured. We typically use `echo "" | gzip > file.gz` instead.

File: modules/nf-core/nanoq/tests/main.nf.test.snap [hunk 2]
```diff
@@ -0,0 +1,307 @@
...
+                "reads": [
+                    [
+                        {
+                            "id": "test",
+                            "single_end": true
+                        },
+                        [
+                            "test.stats:md5,5ab32af3352dfeca8268e10edf6e4dbe",
+                            "test_nanoq.fastq.bz2:md5,b53cf14fd4eb5b16c459c41f03cc8a4b",
+                            "versions.yml:md5,7a40efe417ff7dbb9e91e9c1629a04e6"
```
Comment (SPPearce) [inline modules/nf-core/nanoq/tests/main.nf.test.snap:100]: This looks like your expression for the output channel is capturing all these.

File: modules/nf-core/nanoq/main.nf [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+        'https://depot.galaxyproject.org/singularity/nanoq:0.10.0--h031d066_2' :
+        'biocontainers/nanoq:0.10.0--h031d066_2'}"
+
+    input:
+    tuple val(meta), path(ontreads)
+    val(output_type) // u: uncompressed; b: Bzip2; g: Gzip; l: Lzma
+
+    output:
+    tuple val(meta), path("*.stats")                                                  , emit: stats
+    tuple val(meta), path("*")                                                        , emit: reads
```
Comment (SPPearce) [inline modules/nf-core/nanoq/main.nf:18]: Yes, this is too broad and capturing the stats and versions as well.

File: modules/nf-core/nanoq/main.nf [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+
+    conda "${moduleDir}/environment.yml"
+
+    container "${workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
+        'https://depot.galaxyproject.org/singularity/nanoq:0.10.0--h031d066_2' :
+        'biocontainers/nanoq:0.10.0--h031d066_2'}"
+
+    input:
+    tuple val(meta), path(ontreads)
+    val(output_type) // u: uncompressed; b: Bzip2; g: Gzip; l: Lzma
```
[the remark this answers is not in the document]
Comment (LilyAnderssonLee) [reply modules/nf-core/nanoq/main.nf:12]: Right, that's why all three files were captured by content 1.

File: modules/nf-core/nanoq/main.nf (hunk 3 above)
Comment (LilyAnderssonLee) [reply modules/nf-core/nanoq/main.nf:16]: Right, that's why all three files were captured by content 1.

File: modules/nf-core/nanoq/main.nf [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+    tuple val(meta), path("*")                                                        , emit: reads
+    path "versions.yml"                                                               , emit: versions
+
+    when:
+    task.ext.when == null || task.ext.when
+
+    script:
+    def args = task.ext.args ?: ''
+    def prefix = task.ext.prefix ?: "${meta.id}"
+    def seq_type = ontreads.baseName.split("\\.")[1]
```
[the remark this answers is not in the document]
Comment (SPPearce) [reply modules/nf-core/nanoq/main.nf:25]: What are the actual options?

File: modules/nf-core/nanoq/tests/main.nf.test.snap (hunk 1 above)
Comment (LilyAnderssonLee) [reply modules/nf-core/nanoq/tests/main.nf.test.snap]: `nanoq` can directly handle the compression.

File: modules/nf-core/nanoq/tests/main.nf.test.snap (hunk 2 above)
Comment (LilyAnderssonLee) [reply modules/nf-core/nanoq/tests/main.nf.test.snap]: Updated.

File: modules/nf-core/nanoq/main.nf (hunk 4 above)
[the remark this answers is not in the document]
Comment (LilyAnderssonLee) [reply modules/nf-core/nanoq/main.nf]: Updated.

File: modules/nf-core/nanoq/main.nf (hunk 5 above)
[the remark this answers is not in the document]
Comment (LilyAnderssonLee) [reply modules/nf-core/nanoq/main.nf]: Any fasta (fa, fna, faa,fasta) or fastq (fq,fastq) formats.

File: modules/nf-core/nanoq/main.nf [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+        'https://depot.galaxyproject.org/singularity/nanoq:0.10.0--h031d066_2' :
+        'biocontainers/nanoq:0.10.0--h031d066_2'}"
+
+    input:
+    tuple val(meta), path(ontreads)
+    val(output_format) //One of the following: fastq, fastq.gz, fastq.bz2, fastq.lzma, fasta, fasta.gz, fasta.bz2, fasta.lzma.
+
+    output:
+    tuple val(meta), path("*.stats")                                                  , emit: stats
+    tuple val(meta), path("*_nanoq.${output_format}")                                                 , emit: reads
```
Comment (SPPearce) [inline modules/nf-core/nanoq/main.nf:16]: Can you line the emits up and push them back to the left please.

File: modules/nf-core/nanoq/main.nf [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+    path "versions.yml"                                                               , emit: versions
+
+    when:
+    task.ext.when == null || task.ext.when
+
+    script:
+    def args = task.ext.args ?: ''
+    def prefix = task.ext.prefix ?: "${meta.id}"
```
Comment (SPPearce) [inline modules/nf-core/nanoq/main.nf:24]:
```suggestion
    def prefix = task.ext.prefix ?: "${meta.id}_nanoq"
```

File: modules/nf-core/nanoq/main.nf [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+
+    """
+    nanoq -i $ontreads \\
+        -r ${prefix}.stats \\
+        -o ${prefix}_nanoq.$output_format
```
Comment (SPPearce) [inline modules/nf-core/nanoq/main.nf:29]:
```suggestion
        -o ${prefix}.$output_format
```

File: modules/nf-core/nanoq/main.nf [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+        nanoq: \$(nanoq --version | sed -e 's/nanoq //g')
+    END_VERSIONS
+    """
+
+    stub:
+    def args = task.ext.args ?: ''
+    def prefix = task.ext.prefix ?: "${meta.id}"
+
+    """
+    touch ${prefix}_nanoq.$output_format
```
Comment (SPPearce) [inline modules/nf-core/nanoq/main.nf:42]:
```suggestion
    echo "" | gzip > ${prefix}_nanoq.$output_format
```

File: modules/nf-core/nanoq/main.nf (hunk 6 above)
Comment (SPPearce) [inline modules/nf-core/nanoq/main.nf:16]: Are these "reads" if they are fasta sequences?

File: modules/nf-core/nanoq/tests/main.nf.test.snap (hunk 1 above)
Comment (SPPearce) [reply modules/nf-core/nanoq/tests/main.nf.test.snap]: This issue is just in the stub, where it tries to go inside .gz files and can't manage it. This causes for some reason a full path to appear in the snapshot, which is causing your failures in the CI/CD.

File: modules/nf-core/nanoq/main.nf (hunk 6 above)
Comment (LilyAnderssonLee) [reply modules/nf-core/nanoq/main.nf:16]: Yes, those reads could be `fasta` format.

Comment (SPPearce) [reply modules/nf-core/nanoq/main.nf]: I just meant whether "reads" was a good name, as I assume that to be fastq files (but if it is fine that's not a problem, just asking the quesiton)

File: modules/nf-core/nanoq/main.nf [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+    cat <<-END_VERSIONS > versions.yml
+    "${task.process}":
+        nanoq: \$(nanoq --version | sed -e 's/nanoq //g')
+    END_VERSIONS
+    """
+
+    stub:
+    def args = task.ext.args ?: ''
+    def prefix = task.ext.prefix ?: "${meta.id}_nanoq"
+
```
Comment (SPPearce) [inline modules/nf-core/nanoq/main.nf:40]:
```suggestion
```

File: modules/nf-core/nanoq/main.nf [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+    tuple val(meta), path("*_nanoq.${output_format}")                                 , emit: reads
+    path "versions.yml"                                                               , emit: versions
+
+    when:
+    task.ext.when == null || task.ext.when
+
+    script:
+    def args = task.ext.args ?: ''
+    def prefix = task.ext.prefix ?: "${meta.id}_nanoq"
+
```
Comment (SPPearce) [inline modules/nf-core/nanoq/main.nf:25]:
```suggestion
```

**Review by SPPearce — APPROVED** [review]
Looks good to me now.

**Comment by LilyAnderssonLee** [comment]
@SPPearce Thanks for your help!!!

**Pull request merged**